pkg/controller/kms/key/setup.go

```diff
@@ -0,0 +1,330 @@
+			KeyId: aws.String(meta.GetExternalName(cr)),
+		}); err != nil {
+			return err
+		}
+	} else {
+		if _, err := u.client.DisableKeyWithContext(ctx, &svcsdk.DisableKeyInput{
+			KeyId: aws.String(meta.GetExternalName(cr)),
+		}); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+type deleter struct {
+	client svcsdkapi.KMSAPI
+}
+
+// schedule for deletion instead of delete
+func (d *deleter) delete(ctx context.Context, mg resource.Managed) error {
+	cr, ok := mg.(*svcapitypes.Key)
+	if !ok {
+		return errors.New(errUnexpectedObject)
+	}
+
+	req := &svcsdk.ScheduleKeyDeletionInput{
+		KeyId: aws.String(meta.GetExternalName(cr)),
+	}
+
+	if cr.Spec.ForProvider.PendingWindowInDays != nil {
+		req.PendingWindowInDays = aws.Int64(*cr.Spec.ForProvider.PendingWindowInDays)
+	}
+
+	_, err := d.client.ScheduleKeyDeletionWithContext(ctx, req)
+
+	return errors.Wrap(err, errDelete)
+}
+
+type observer struct {
+	client svcsdkapi.KMSAPI
+}
+
+func (o *observer) lateInitialize(in *svcapitypes.KeyParameters, obj *svcsdk.DescribeKeyOutput) error {
+	// Policy
+	if in.Policy == nil {
+		resPolicy, err := o.client.GetKeyPolicy(&svcsdk.GetKeyPolicyInput{
+			KeyId:      obj.KeyMetadata.KeyId,
+			PolicyName: aws.String("default"),
+		})
+
+		if err != nil {
+			return err
+		}
+
+		if resPolicy.Policy != nil {
+			in.Policy = resPolicy.Policy
+		}
+	}
+
+	// Enabled
+	if in.Enabled == nil {
+		in.Enabled = obj.KeyMetadata.Enabled
+	}
+
+	if len(in.Tags) == 0 {
+		resTags, err := o.client.ListResourceTags(&svcsdk.ListResourceTagsInput{
+			KeyId: obj.KeyMetadata.KeyId,
+		})
+
+		if err != nil {
+			return err
+		}
+
+		if len(resTags.Tags) > 0 {
+			for _, t := range resTags.Tags {
+				in.Tags = append(in.Tags, &svcapitypes.Tag{
+					TagKey:   t.TagKey,
+					TagValue: t.TagValue,
+				})
+			}
+		}
+	}
+
+	return nil
+}
+
+func (o *observer) isUpToDate(cr *svcapitypes.Key, obj *svcsdk.DescribeKeyOutput) (bool, error) {
+	// Description
+	if obj.KeyMetadata.Description != nil &&
+		cr.Spec.ForProvider.Description != nil &&
+		*obj.KeyMetadata.Description != *cr.Spec.ForProvider.Description {
+		return false, nil
+	}
+
+	// KeyPolicy
+	resPolicy, err := o.client.GetKeyPolicy(&svcsdk.GetKeyPolicyInput{
+		KeyId:      aws.String(meta.GetExternalName(cr)),
+		PolicyName: aws.String("default"),
+	})
+	if err != nil {
+		return false, err
+	}
+	if *cr.Spec.ForProvider.Policy != *resPolicy.Policy {
+		return false, nil
+	}
+
+	// Tags
+	resTags, err := o.client.ListResourceTags(&svcsdk.ListResourceTagsInput{
+		KeyId: aws.String(meta.GetExternalName(cr)),
+	})
+	if err != nil {
+		return false, nil
+	}
+	return compareTags(cr.Spec.ForProvider.Tags, resTags.Tags), nil
+}
+
+// returns which AWS Tags exist in the resource tags and which are outdated and should be removed
+func diffTags(tags []*svcapitypes.Tag, kmsTags []*svcsdk.Tag) ([]*svcapitypes.Tag, []*svcsdk.Tag) {
+	crTags := make(map[string]*svcapitypes.Tag)
```

Comment:
Can we use an implementation similar to the one [here](https://github.com/crossplane/provider-aws/pull/529/files#diff-4c299908ebe9a0955fdcc9ea1c4e423253b0f582ef5a70e1fd0f1e9328ddd574R159) so that we don't need to have any loops/logic outside of `diffTags` where it's called.